pkg/registry/apis/provisioning/resources/idmapper.go

```diff
@@ -0,0 +1,75 @@
+package resources
+
+import (
+	"crypto/sha256"
+	"encoding/base64"
+	"path"
+	"strings"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/grafana/grafana/pkg/apimachinery/utils"
+	"github.com/grafana/grafana/pkg/infra/slugify"
+)
+
+// Get a filename for a given object
+type FileNameMapper func(obj metav1.Object) string
+
+// Get the object/folder names from a path+object
+type NameMapper func(repo string, path string, obj metav1.Object) (objectName string, folderName string)
+
+// Uses the k8s uid name directly
+func FileNameFromK8sName(obj metav1.Object) string {
+	return obj.GetName() // sanitize?
+}
+
+// Calculates a human readable file name from a title field
+func FileNameFromSlug(obj metav1.Object) string {
+	meta, err := utils.MetaAccessor(obj)
+	if err == nil {
+		title := meta.FindTitle("")
+		if title != "" {
+			return slugify.Slugify(title)
+		}
+	}
+	return FileNameFromK8sName(obj)
+}
+
+// This picks the names from the saved the metadata
+func NamesFromMetadata(_ string, _ string, obj metav1.Object) (objectName string, folderName string) {
+	objectName = obj.GetName()
+	anno := obj.GetAnnotations()
+	if anno != nil {
+		folderName = anno[utils.AnnoKeyFolder]
+	}
+	return
+}
+
+// Will pick a name based on the hashed repository and path
+func NamesFromHashedRepoPath(repo string, fpath string, _ metav1.Object) (objectName string, folderName string) {
+	// remove the extension
```

Comment:
this implies we should error if a ".yaml" and ".json" exist with the same prefix 🤔 

alternatively we keep the extension in the hash, but do not display it.  Means changing from json/yaml is a new resoruce, but I think that is OK